File: backend/employer-profile-service/src/main/java/com/jobportal/employerprofile/service/EmployerProfileService.java
package com.jobportal.employerprofile.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import com.jobportal.employerprofile.dto.EmployerProfileRequestDto;
import com.jobportal.employerprofile.dto.EmployerProfileResponseDto;
import com.jobportal.employerprofile.entities.EmployerProfile;
import com.jobportal.employerprofile.repository.EmployerProfileRepository;



@Service
public class EmployerProfileService {

    @Autowired
    private EmployerProfileRepository employerProfileRepository;

    @Transactional
    public EmployerProfileResponseDto createProfile(EmployerProfileRequestDto request) {
        if (employerProfileRepository.existsByUserId(request.getUserId())) {
            throw new ResponseStatusException(HttpStatus.CONFLICT,
                    "An employer profile already exists for this user");
        }
        if (employerProfileRepository.existsByRegistrationId(request.getRegistrationId())) {
            throw new ResponseStatusException(HttpStatus.CONFLICT, "Registration id is already in use");
        }

        EmployerProfile employerProfile = new EmployerProfile();
        copyRequest(request, employerProfile);
        return toResponse(employerProfileRepository.save(employerProfile));
    }

    @Transactional(readOnly = true)
    public EmployerProfileResponseDto getById(Integer employerId) {
        return toResponse(findEmployer(employerId));
    }

    @Transactional(readOnly = true)
    public EmployerProfileResponseDto getByUserId(Integer userId) {
        return employerProfileRepository.findByUserId(userId)
                .map(this::toResponse)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND,
                        "Employer profile not found for user id: " + userId));
    }

    @Transactional
    public EmployerProfileResponseDto updateProfile(Integer employerId, EmployerProfileRequestDto request) {
        EmployerProfile employerProfile = findEmployer(employerId);

        if (!employerProfile.getUserId().equals(request.getUserId())) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "User id cannot be changed");
        }

        if (!employerProfile.getRegistrationId().equals(request.getRegistrationId())
                && employerProfileRepository.existsByRegistrationId(request.getRegistrationId())) {
            throw new ResponseStatusException(HttpStatus.CONFLICT, "Registration id is already in use");
        }

        copyRequest(request, employerProfile);
        return toResponse(employerProfileRepository.save(employerProfile));
    }

    @Transactional
    public void deleteProfile(Integer employerId) {
        employerProfileRepository.delete(findEmployer(employerId));
    }

    private EmployerProfile findEmployer(Integer employerId) {
        return employerProfileRepository.findById(employerId)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND,
                        "Employer profile not found with id: " + employerId));
    }

    private void copyRequest(EmployerProfileRequestDto request, EmployerProfile employerProfile) {
        employerProfile.setUserId(request.getUserId());
        employerProfile.setCompanyName(request.getCompanyName().trim());
        employerProfile.setEmail(request.getEmail().trim().toLowerCase());
        employerProfile.setAddress(request.getAddress());
        employerProfile.setCity(request.getCity());
        employerProfile.setState(request.getState());
        employerProfile.setCountry(request.getCountry().trim());
        employerProfile.setRegistrationId(request.getRegistrationId().trim());
        employerProfile.setDescription(request.getDescription());
        employerProfile.setIndustry(request.getIndustry().trim());
    }

    private EmployerProfileResponseDto toResponse(EmployerProfile employerProfile) {
        EmployerProfileResponseDto response = new EmployerProfileResponseDto();
        response.setEmployerId(employerProfile.getEmployerId());
        response.setUserId(employerProfile.getUserId());
        response.setCompanyName(employerProfile.getCompanyName());
        response.setEmail(employerProfile.getEmail());
        response.setAddress(employerProfile.getAddress());
        response.setCity(employerProfile.getCity());
        response.setState(employerProfile.getState());
        response.setCountry(employerProfile.getCountry());
        response.setRegistrationId(employerProfile.getRegistrationId());
        response.setDescription(employerProfile.getDescription());
        response.setIndustry(employerProfile.getIndustry());
        return response;
    }
}

File: backend-dotnet/Services/JobPortal.EmployerProfileService/Services/EmployerProfileService.cs
using AutoMapper;
using System.Net;
using JobPortal.EmployerProfiles.DTOs;
using JobPortal.EmployerProfiles.Exceptions;
using JobPortal.EmployerProfiles.Interfaces;
using JobPortal.EmployerProfiles.Models;

namespace JobPortal.EmployerProfiles.Services;

public class EmployerProfileService
{
    private readonly IEmployerProfileRepository _repository;
    private readonly IMapper _mapper;

    public EmployerProfileService(IEmployerProfileRepository repository, IMapper mapper)
    {
        _repository = repository;
        _mapper = mapper;
    }

    public async Task<EmployerProfileResponseDto> CreateProfileAsync(EmployerProfileRequestDto request)
    {
        if (request.UserId.HasValue && await _repository.ExistsByUserIdAsync(request.UserId.Value))
        {
            throw new HttpStatusException(HttpStatusCode.Conflict, "An employer profile already exists for this user");
        }

        if (await _repository.ExistsByRegistrationIdAsync(request.RegistrationId))
        {
            throw new HttpStatusException(HttpStatusCode.Conflict, "Registration id is already in use");
        }

        var employerProfile = _mapper.Map<EmployerProfile>(request);
        var savedProfile = await _repository.SaveAsync(employerProfile);
        return _mapper.Map<EmployerProfileResponseDto>(savedProfile);
    }

    public async Task<EmployerProfileResponseDto> GetByIdAsync(int employerId)
    {
        var profile = await _repository.GetByIdAsync(employerId);
        if (profile == null)
        {
            throw new HttpStatusException(HttpStatusCode.NotFound, $"Employer profile not found with id: {employerId}");
        }
        return _mapper.Map<EmployerProfileResponseDto>(profile);
    }

    public async Task<List<EmployerProfileResponseDto>> GetAllProfilesAsync()
    {
        var profiles = await _repository.GetAllAsync();
        return _mapper.Map<List<EmployerProfileResponseDto>>(profiles);
    }

    public async Task<EmployerProfileResponseDto> GetByUserIdAsync(int userId)
    {
        var profile = await _repository.GetByUserIdAsync(userId);
        if (profile == null)
        {
            throw new HttpStatusException(HttpStatusCode.NotFound, $"Employer profile not found for user id: {userId}");
        }
        return _mapper.Map<EmployerProfileResponseDto>(profile);
    }

    public async Task<EmployerProfileResponseDto> UpdateProfileAsync(int employerId, EmployerProfileRequestDto request)
    {
        var profile = await _repository.GetByIdAsync(employerId);
        if (profile == null)
        {
            throw new HttpStatusException(HttpStatusCode.NotFound, $"Employer profile not found with id: {employerId}");
        }

        if (profile.UserId != request.UserId)
        {
            throw new HttpStatusException(HttpStatusCode.BadRequest, "User id cannot be changed");
        }

        if (profile.RegistrationId != request.RegistrationId && await _repository.ExistsByRegistrationIdAsync(request.RegistrationId))
        {
            throw new HttpStatusException(HttpStatusCode.Conflict, "Registration id is already in use");
        }

        _mapper.Map(request, profile);
        var savedProfile = await _repository.SaveAsync(profile);
        return _mapper.Map<EmployerProfileResponseDto>(savedProfile);
    }

    public async Task DeleteProfileAsync(int employerId)
    {
        var profile = await _repository.GetByIdAsync(employerId);
        if (profile == null)
        {
            throw new HttpStatusException(HttpStatusCode.NotFound, $"Employer profile not found with id: {employerId}");
        }
        await _repository.DeleteAsync(profile);
    }
}
